clarity/utils/source_separation.py

```diff
@@ -0,0 +1,101 @@
+import torch
+import torchaudio
+from torchaudio.pipelines import HDEMUCS_HIGH_MUSDB_PLUS
+from torchaudio.transforms import Fade
+
+
+def source_separation(mix, model, sample_rate, segment=10.0, overlap=0.1, device=None):
```

Comment:
Is it possible to devise a unit-test for this function? Ideally parameterised so that the conditions of different values for `mix.ndim` and `mix.shape` are covered.

Generally there is quite a lot of steps undertaken in this function though and it would be good to break each step out into its own very small function that then has single responsibility and unit-tests can then be written for each.

For example `# Split audio into segments` could be a function on its own, as could `# Move model and audio to device` and `separated_audio = torch.zeros(...` and a few others. If all these things "hang together" then they could be grouped into a Class.

I'll have a look and feedback more tomorrow.

Reply:
Good point, thank you
I just created one test to submit the PR and discuss the code and the location where to save it.
Will improve the test following your recommendations.